dep0/lib/02_parser/src/parse.cpp

```diff
@@ -120,38 +120,58 @@ struct parse_visitor_t : dep0::DepCParserVisitor
             get_loc(src, *ctx).value(),
-            std::any_cast<type_t>(visitType(ctx->type())),
             get_text(src, *ctx->name).value(),
-            fmap(ctx->arg(), [this] (auto* x) { return std::any_cast<func_def_t::arg_t>(visitArg(x)); }),
-            std::any_cast<body_t>(visitBody(ctx->body()))
-        };
+            expr_t::abs_t{
+                fmap(ctx->arg(), [this] (auto* x) { return std::any_cast<expr_t::abs_t::arg_t>(visitArg(x)); }),
+                std::any_cast<type_t>(visitType(ctx->type())),
+                std::any_cast<body_t>(visitBody(ctx->body()))
+            }};
     }
 
     virtual std::any visitType(DepCParser::TypeContext* ctx) override
     {
-        if (ctx->KW_BOOL()) return type_t{get_loc(src, *ctx).value(), type_t::bool_t{}};
-        if (ctx->KW_UNIT_T()) return type_t{get_loc(src, *ctx).value(), type_t::unit_t{}};
-        if (ctx->KW_I8_T()) return type_t{get_loc(src, *ctx).value(), type_t::i8_t{}};
-        if (ctx->KW_I16_T()) return type_t{get_loc(src, *ctx).value(), type_t::i16_t{}};
-        if (ctx->KW_I32_T()) return type_t{get_loc(src, *ctx).value(), type_t::i32_t{}};
-        if (ctx->KW_I64_T()) return type_t{get_loc(src, *ctx).value(), type_t::i64_t{}};
-        if (ctx->KW_U8_T()) return type_t{get_loc(src, *ctx).value(), type_t::u8_t{}};
-        if (ctx->KW_U16_T()) return type_t{get_loc(src, *ctx).value(), type_t::u16_t{}};
-        if (ctx->KW_U32_T()) return type_t{get_loc(src, *ctx).value(), type_t::u32_t{}};
-        if (ctx->KW_U64_T()) return type_t{get_loc(src, *ctx).value(), type_t::u64_t{}};
-        if (ctx->name) return type_t{get_loc(src, *ctx).value(), type_t::name_t{get_text(src, *ctx->name).value()}};
-        assert(nullptr);
+        assert(ctx);
+        auto const loc = get_loc(src, *ctx).value();
+        if (ctx->KW_BOOL()) return type_t{loc, type_t::bool_t{}};
+        if (ctx->KW_UNIT_T()) return type_t{loc, type_t::unit_t{}};
+        if (ctx->KW_I8_T()) return type_t{loc, type_t::i8_t{}};
+        if (ctx->KW_I16_T()) return type_t{loc, type_t::i16_t{}};
+        if (ctx->KW_I32_T()) return type_t{loc, type_t::i32_t{}};
+        if (ctx->KW_I64_T()) return type_t{loc, type_t::i64_t{}};
+        if (ctx->KW_U8_T()) return type_t{loc, type_t::u8_t{}};
+        if (ctx->KW_U16_T()) return type_t{loc, type_t::u16_t{}};
+        if (ctx->KW_U32_T()) return type_t{loc, type_t::u32_t{}};
+        if (ctx->KW_U64_T()) return type_t{loc, type_t::u64_t{}};
+        if (auto const types = ctx->argType(); not types.empty())
+            return type_t{
+                loc,
+                type_t::arr_t{
+                    fmap(types, [this] (auto* x) { return std::any_cast<type_t::arr_t::arg_kind_t>(visitArgType(x)); }),
+                    std::any_cast<type_t>(visitType(ctx->type()))}};
+        if (ctx->name) return type_t{loc, type_t::var_t{get_text(src, *ctx->name).value()}};
+        throw error_t{"unexpected alternative when parsing TypeContext", loc};
```

Comment:
should mark the public parse function as noexcept since we are catching this before return from there